ledger/internal/eval.go

```diff
@@ -1564,29 +1559,67 @@ func Eval(ctx context.Context, l LedgerForEvaluator, blk bookkeeping.Block, vali
+		case txgroup, ok := <-paysetgroupsPreloadedResourcesCh:
 			if !ok {
 				break transactionGroupLoop
 			} else if txgroup.err != nil {
 				return ledgercore.StateDelta{}, txgroup.err
 			}
 
-			for _, br := range txgroup.balances {
-				base.accounts[br.Addr] = br.AccountData
-				// TODO: pull needed resources into cache?
+			for _, br := range txgroup.accounts {
+				if _, have := base.accounts[*br.address]; !have {
+					base.accounts[*br.address] = *br.data
+				}
 			}
-			err = eval.TransactionGroup(txgroup.group)
+			for _, lr := range txgroup.resources {
+				if lr.address == nil {
+					// we attempted to look for the creator, and failed.
+					if lr.creatableType == basics.AssetCreatable {
+						base.creators[creatable{cindex: lr.creatableIndex, ctype: basics.AssetCreatable}] = foundAddress{exists: false}
+					} else {
+						base.creators[creatable{cindex: lr.creatableIndex, ctype: basics.AppCreatable}] = foundAddress{exists: false}
+					}
+					continue
+				}
+				if lr.creatableType == basics.AssetCreatable {
+					if lr.resource.AssetHolding != nil {
+						base.assets[ledgercore.AccountAsset{Address: *lr.address, Asset: basics.AssetIndex(lr.creatableIndex)}] = cachedAssetHolding{value: *lr.resource.AssetHolding, exists: true}
+					} else {
+						base.assets[ledgercore.AccountAsset{Address: *lr.address, Asset: basics.AssetIndex(lr.creatableIndex)}] = cachedAssetHolding{exists: false}
+					}
+					if lr.resource.AssetParams != nil {
+						base.assetParams[ledgercore.AccountAsset{Address: *lr.address, Asset: basics.AssetIndex(lr.creatableIndex)}] = cachedAssetParams{value: *lr.resource.AssetParams, exists: true}
+						base.creators[creatable{cindex: lr.creatableIndex, ctype: basics.AssetCreatable}] = foundAddress{address: *lr.address, exists: true}
+					} else {
+						base.assetParams[ledgercore.AccountAsset{Address: *lr.address, Asset: basics.AssetIndex(lr.creatableIndex)}] = cachedAssetParams{exists: false}
+
+					}
+				} else {
+					if lr.resource.AppLocalState != nil {
+						base.appLocalStates[ledgercore.AccountApp{Address: *lr.address, App: basics.AppIndex(lr.creatableIndex)}] = cachedAppLocalState{value: *lr.resource.AppLocalState, exists: true}
+					} else {
+						base.appLocalStates[ledgercore.AccountApp{Address: *lr.address, App: basics.AppIndex(lr.creatableIndex)}] = cachedAppLocalState{exists: false}
+					}
+					if lr.resource.AppParams != nil {
+						base.appParams[ledgercore.AccountApp{Address: *lr.address, App: basics.AppIndex(lr.creatableIndex)}] = cachedAppParams{value: *lr.resource.AppParams, exists: true}
+						base.creators[creatable{cindex: lr.creatableIndex, ctype: basics.AppCreatable}] = foundAddress{address: *lr.address, exists: true}
+					} else {
+						base.appParams[ledgercore.AccountApp{Address: *lr.address, App: basics.AppIndex(lr.creatableIndex)}] = cachedAppParams{exists: false}
+					}
+				}
+			}
+			err = eval.TransactionGroup(paysetgroups[txnGroupIndex])
```

Comment:
I was thinking about that - since that what we had before. However, the transaction are always being executed at the exact same order as above. Adding a slice to the data passed back ( or even an index ), is just a redundant piece of data, since we don't really need it. ( it does make it more "encapsulated", sure )

Reply:
I still think it is better to avoid external indexing to have a bit better code

Reply:
agreed, it makes it harder to follow, I would argue for putting the group slice back into loadedTransactionGroup.

Reply:
that already have been addressed..